Add support for "Type T: Task Manged" activities

mmoore99 opened the pull request:
Fixes #507

This PR implements the following:

- The ability to specify whether to limit the number of volunteers to the number required.  See my comment in #504 for more a detailed description. This capability is in place for both Activities and Tasks.  The "Wait List" capability has not yet been implemented.

- Functionality changes on the Activity Detail page for a  "Task Managed" activity:
  - The "My Assigned Tasks" list was converted to an "accordion" .  Expanding the accordion provides additional information for the task, including: task description, number of users required and available spots, and skills required. The expanded panel also contains buttons for the various actions that can be performed.

  - An additional "Tasks" list, also an accordion, was created that contains all tasks for the activity that the user is not signed up for.  The user can sign up for tasks from this list. A filter option has been provided to hide tasks that are full. If `IsLimitVolunteers` is true for a task, the task name for tasks that are not full is shown in bold.

  - All of the actions, including: volunteering, accepting a task assignment, rejecting a task assignment, marking a task as complete, marking a task as unable to complete and unenrolling from a task have been converted to Ajax calls and no page refreshes occur for these actions.

  - Confirmation messages are now displayed after completion of each action.

- When a user volunteers for a task an `ActivitySignup` is created if one does not already exist for the activity.  when a user unenrolls from a task the `ActivitySignup` is deleted if the user is not signed up for any other tasks in the activity.

![screenclip](https://cloud.githubusercontent.com/assets/45022/13467896/2169650c-e055-11e5-95b4-615709b5c740.png)

![screenclip 1](https://cloud.githubusercontent.com/assets/45022/13467900/279f1304-e055-11e5-8c1d-fcc189035902.png)

![screenclip 2](https://cloud.githubusercontent.com/assets/45022/13467906/304c7154-e055-11e5-83d2-0f3b246c7ad6.png)

![screenclip 3](https://cloud.githubusercontent.com/assets/45022/13467909/357a25fe-e055-11e5-9012-457690b3e53b.png)

![screenclip 4](https://cloud.githubusercontent.com/assets/45022/13467913/3ae9e682-e055-11e5-8402-0b817ecba065.png)

![screenclip 5](https://cloud.githubusercontent.com/assets/45022/13467919/3fae389e-e055-11e5-92df-e470836966e4.png)

![screenclip 6](https://cloud.githubusercontent.com/assets/45022/13467934/48d6e24a-e055-11e5-96ef-8dee54319715.png)

![screenclip 9](https://cloud.githubusercontent.com/assets/45022/13166262/7c6475f8-d67a-11e5-8e11-d49b52d670b5.png)

tonysurma commented:
last one on the stack.  After we know what to do with the others I will do same here.  Can someone also review this as the others have a review already in place?  Thanks.

mmoore99 commented:
### WIP - Changes made during the review process of #518 will need to be reflected and reconciled with this PR.  Please hold off on review of this PR until those changes have been made and this PR has been rebased.

tonysurma commented:
@mmoore99   please rebase off of master when ready as we now have a fix for the build/test issues!

For reference on rebasing this page helps (replace upstream with HTBox in examples) https://robots.thoughtbot.com/keeping-a-github-fork-updated

mmoore99 commented:
@tonysurma @mheggeseth I believe this PR is now ready for review.  I have updated the screen shots in the initial comment to reflect UI changes that have been previously discussed.  The branch has been rebased off of master.

tonysurma commented:
Thanks @mmoore99, it looks like you will need (yet another :)) rebase
@mheggeseth @MisterJames can you review

tonysurma commented:
Thanks @mmoore99, it looks like you will need (yet another :)) rebase
@mheggeseth @MisterJames can you review

mmoore99 commented:
@tonysurma rebase completed

mmoore99 commented:
@tonysurma another rebase completed

tonysurma commented:
Thx @mmoore99, once we get a pr review I will pull it in.  kindly pinging @MisterJames @mheggeseth @dpaquette :smiley_cat:

mheggeseth commented:
@tonysurma @mmoore99 I can review (at least some) this weekend. There's a lot here.

MisterJames commented:
I'll start reviewing this one now @mheggeseth

MisterJames commented on AllReadyApp/Web-App/AllReady/Areas/Admin/Models/TaskSummaryModel.cs:
```diff
@@ -36,5 +37,6 @@ public class TaskSummaryModel
         public List<VolunteerModel> AllVolunteers { get; set; } = new List<VolunteerModel>();
 
         public int AcceptedVolunteerCount => AssignedVolunteers?.Where(v => v.HasVolunteered).Count() ?? 0;
+        public int NumberOfVolunteersSignedUp { get; set; }
```
Is this different than `AllVolunteers.Count()`? How is this set/updated differently?

mmoore99 replied:
`AllVolunteers` refers to volunteers for the activity, not the task. `AssignedVolunteers` refers to volunteers for the task.  At the moment I cannot recall why I did not use `AssignedVolunteers.Count`, so I have changed it to that for now.

MisterJames commented on AllReadyApp/Web-App/AllReady/Controllers/TaskApiController.cs:
```diff
@@ -131,33 +148,74 @@ private bool HasTaskSignupEditPermissions(AllReadyTask task)
-        public async void Signup(int taskId, string userId = "")
-        {
-            var task = _allReadyDataAccess.GetTask(taskId);
+        //[HttpGet]
```
If this code block is dead, we can delete it.

mmoore99 replied:
Done.

MisterJames commented on AllReadyApp/Web-App/AllReady/Controllers/TaskApiController.cs:
```diff
@@ -1,23 +1,40 @@
+
 
-        public TaskApiController(IAllReadyDataAccess allReadyDataAccess)
+        public TaskApiController(IAllReadyDataAccess allReadyDataAccess, IMediator bus)
```
Just to be consistent...we started by naming `bus` originally, but have switched over to `mediator`. Could you rename this guy?

mmoore99 replied:
done

MisterJames commented on AllReadyApp/Web-App/AllReady/Features/Notifications/NotifyAdminForUserUnenrolls.cs:
```diff
@@ -22,58 +24,93 @@ public NotifyAdminForUserUnenrolls(IMediator mediator, IOptions<GeneralSettings>
+            }
+            catch (Exception e)
+            {
+                // ignored
```
I see what you're doing here, but this makes it hard to troubleshoot :(

Can you perhaps inject something like an `ILogger<NotifyAdminForUserUnenrolls> logger` into the constructor and log out the exception? Logging to the console is configured for the project, so at least if we had a debugger attached (or we were checking the Azure logs) we'd see the problem with notifications.

mmoore99 replied:
Done.  I agree that is a much better approach.

MisterJames commented:
@mmoore99  this was a great PR and you bit off a big chunk of work here. It is much appreciated. I've requested a few small code cleanups, but otherwise:
 - Tests were included
 - DB migrations worked locally for me
 - Unit tests passing locally (and on build server)
 - Application runs
 - From what I believe I understand of the PR and original issue from the request and conversation(s) :eyes: it seems like most of the check boxes were ticked (save the wait list) :)

For future PRs I have the following suggestions: 
 - a couple of the methods in handlers required some DB lookups that were multiple lines (because of the `.ThenInclude`s). It would be great to break out a private method for each of those so that the handler proper is a little easier to read, something like `var thingy = GetThingy(id)`.
 - when you're doing validation in handlers, you could write a `FooValidator` object with public methods that would allow for unit tests on those validations

Finally, I know this was a monumental task, and you broke it out of another monumental task, but I feel like this could have been 4 or 5 seperate PRs. This is a hard thing to ask of a volunteer (especially as a fellow volunteer!) but this one was difficult to evaluate because of the size. Thankfully things were working and your code was at a good level, but when they're this big I can't review them on my lunch hour, if you get what I mean?

Thanks again...I'll merge after your code touches, please ping me.

MisterJames commented:
@mmoore99 You're also going to have a small rebase to tackle because of #596, but it should only be the one file (ActivityController).

mmoore99 commented:
@MisterJames thanks for the review and the comments.  I have made all of the requested changes and rebased.  I fixed a bug that was introduced in the refactoring of `ActivityController` in PR #596.  Everything has now been pushed. My responses to some of your comments follow:

> a couple of the methods in handlers required some DB lookups that were multiple lines (because of the .ThenIncludes). It would be great to break out a private method for each of those so that the handler proper is a little easier to read, something like var thingy = GetThingy(id)

I refactored a number of the handlers as per your suggestion. 

> when you're doing validation in handlers, you could write a FooValidator object with public methods that would allow for unit tests on those validations

I am afraid you lost me here. What validation in handlers are you referring to?  Is there an example that you can point me to that might help clarify?

>  but I feel like this could have been 4 or 5 separate PRs.

Got it!  I apologize for the size of the PR.  I will try to keep them more manageable in the future.

MisterJames commented:
Awesome, thanks for the quick update!

So, as an example of where an external validator might come in handy is here:

https://github.com/HTBox/allReady/pull/523/files#diff-5370386cd1e40cd809df10a8a13d6a4bR42

Definitely not required at this point; I'm actually still trying to come up with a recommendation there, given that a validator of any capacity would require access to the database, which either means taking a dependency or dispatching via MediatR, so it's more work. I've trying the inline approach in my open PR (#602) right now, but I feel like it's probably best to remove the coupling.
